### ISSUE TITLE
Feature Request: Allow `OptOut` of singleton approach for `None`

### ISSUE THREAD

**Issue by MaSpeng** [schmittjoh/php-option#65]
### Request

Allow opting out of the singleton pattern for the `None` class.

### Description

The `None` class currently follows the singleton approach. Due to the ability of [PHP 8.1 to use new initializers](https://stitcher.io/blog/php-81-new-in-initializers) there are scenarios where it would be really useful to actually `OptOut` of the singleton approach.

One example I currently facing is the usage of the `symfony/serializer` component, DTO, and a GraphQL API.

Given the following example:

```GraphQL
mutation {
    example(
        requiredValue: String!
        optionalValue: String
    ): {
        requiredValue
        optionalValue
    }
}
```

```PHP
class Dto {
   public function __construct(
       public readonly string $requiredValue,
       public readonly Option $optionalValue
   ) {
   }
}

$serializer = new Serializer([new ObjectNormalizer()], []);

// This call will fail since no value has been provided for the constructor argument `$optionalValue`
$dto = $serializer->denormalize(
    [
        'requiredValue' => '_value_',
    ],
    Dto::class
);
```

To avoid this error case, the optimal constructor would be:
```diff
class Dto {
   public function __construct(
       public readonly string $requiredValue,
-      public readonly Option $optionalValue
+      public readonly Option $optionalValue = new None()
   ) {
   }
}
```

Sure it would always be possible to handle this before the denormalization process or by providing the list of fallback values to the denormalization process but this would lead to a great amount of duplication and maintenance overhead to a very low cost of runtime and extra memory consumption.

### Acceptance criteria
- Constructor of `None` class is public

**Comment by someniatko** [comment]
IMO it's a deeper PHP problem. IMO PHP itself should not only allow new, but also calling static methods. The reason I think so is "named constructors" pattern got really popular in the PHP community. Ideally it should recognize which are side-effect free and which aren't, but I doubt it will happen. Or, perhaps it should allow marking static methods as being such a constructor, and then allow to call only those.

**Comment by GrahamCampbell** [comment]
Using an option as a parameter type is considered a bad practice. They are intended for return types, only. For this reason, I don't think it makes sense for us to change anything here, since my recommendation would be to replace your optionalValue param with a nullable param. See https://rules.sonarsource.com/java/RSPEC-3553.

**Comment by MaSpeng** [comment]
@someniatko You are right, it would be awesome if we could use the static method right away but PHP does not allow it so, yes we need a workaround to solve this issue which can be removed once a better behavior is in place.

Omitting a workaround because the language does not support it natively makes no sense to me, just think of all those enum libraries out there, their entire use case was to provide functionality that was not yet part of the language itself. The same for psalm which allows marking a class as immutable which can be replaced with read-only attributes and read-only classes in the future.

So personally this is no argument for this issue and should more likely be an RFC for PHP instead.

**Comment by MaSpeng** [comment]
@GrahamCampbell Sadly there is a valid use case to use an option as a parameter type instead of a nullable param.

For example, if you want to use a DTO with a REST API that supports PATCH operations or if you want to use a DTO with GraphQL as a more popular example.

A property in the request could reasonably have 3 states.

A value can be provided to update a state (scalar in this example) or `null` could be provided to reset/clear the state or as an actual value itself, or the property itself is omitted which is not the same as `null` and definitely needs to be distinguished.

Also the reference you have provided is referring to JAVA which has its own language rules which, happily, must not apply to PHP or any other language out there.

So I and surely the other upvoters would really appreciate it if you reconsider opening this issue again as there is no harm in providing this additional functionality. It's not even a BC break.

**Comment by someniatko** [comment]
> Using an option as a parameter type is considered a bad practice. They are intended for return types, only. [...] my recommendation would be to replace your optionalValue param with a nullable param. See https://rules.sonarsource.com/java/RSPEC-3553.

@GrahamCampbell Frankly, the I read the argumentation by following the link, and it is extremely Java-specific. None of the arguments is applicable to PHP - we both don't have method overloading, and PHP does not allow `null` where object is expected. Therefore I still think there is value of using Optional instead of nullable in all contexts (except low-level ones, like DTOs / serialization).

**Comment by GrahamCampbell** [comment]
I still feel like I prefer the singleton approach, anyway. You can still achieve the API you want, but with making the parameter required, with no default.

**Comment by MaSpeng** [comment]
@GrahamCampbell Sure, the most common sense processes to achieve that would be
- create the whole DTO chain on your own
- replicate the knowledge of the request properties, DTO chain, and each DTO's constructor to the serializer context

Booth ways tend to create much overhead, duplication, reduced maintainability, create bound context and increase cognitive load and are also more likely to be error-prone.

It has always to be kept in mind that not all request bodies only contain a single level of properties or simple scalar values.

In this scenario, this is the whole use case of using a serializer to transform your request into a DTO. **D**ont **R**epeat **Yourself** and **K**eep **I**t **Simple**, **S**tupid.

I still have to question, what's the harm with providing the singleton approach and the ability to use the constructor besides one of each of you like more?

**Comment by GrahamCampbell** [comment]
I can't get over the fact that it breaks the comparison against None that we can do today. I would consider changing this in a major version bump, however. It is likely that a major version bump is coming within the next few years.

**Comment by MaSpeng** [comment]
Looking forward to the next major version 😁. Maybe until then, we can use static methods instead of the original constructor.

**Comment by someniatko** [comment]
@GrahamCampbell btw, you've closed this issue "as completed", should be "as wontfix" instead imo.
